File: api/src/main/java/net/asch/bulkit/api/block/state/FilteredIntegerProperty.java
package net.asch.bulkit.api.block.state;

import com.google.common.collect.ImmutableSet;
import net.minecraft.world.level.block.state.properties.Property;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.Optional;

public class FilteredIntegerProperty extends Property<Integer> {
    private final ImmutableSet<Integer> possibleValues;

    protected FilteredIntegerProperty(String name, Collection<Integer> possibleValues) {
        super(name, Integer.class);
        this.possibleValues = ImmutableSet.copyOf(possibleValues);
    }

    protected FilteredIntegerProperty(String name, Integer... possibleValues) {
        super(name, Integer.class);
        this.possibleValues = ImmutableSet.copyOf(possibleValues);
    }

    public @NotNull Collection<Integer> getPossibleValues() {
        return possibleValues;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }

        if (other instanceof FilteredIntegerProperty prop) {
            if (super.equals(other)) {
                return possibleValues.equals(prop.possibleValues);
            }
        }

        return false;
    }

    @Override
    public int generateHashCode() {
        return 31 * super.generateHashCode() + possibleValues.hashCode();
    }

    @Override
    public @NotNull Optional<Integer> getValue(@NotNull String value) {
        try {
            Integer integer = Integer.valueOf(value);
            return possibleValues.contains(integer) ? Optional.of(integer) : Optional.empty();
        } catch (NumberFormatException exc) {
            return Optional.empty();
        }
    }

    @Override
    public @NotNull String getName(@NotNull Integer value) {
        return value.toString();
    }

    public static FilteredIntegerProperty create(String name, Collection<Integer> possibleValues) {
        return new FilteredIntegerProperty(name, possibleValues);
    }

    public static FilteredIntegerProperty create(String name, Integer... possibleValues) {
        return new FilteredIntegerProperty(name, possibleValues);
    }
}

File: api/src/main/kotlin/net/asch/bulkit/api/block/state/FilteredIntegerProperty.kt
package net.asch.bulkit.api.block.state

import com.google.common.collect.ImmutableSet
import net.minecraft.world.level.block.state.properties.Property
import java.util.*

class FilteredIntegerProperty private constructor(name: String, possibleValues: Collection<Int>) :
    Property<Int>(name, Int::class.java) {
    private val possibleValues: ImmutableSet<Int> = ImmutableSet.copyOf(possibleValues)

    private constructor(name: String, vararg possibleValues: Int) : this(name, possibleValues.toList())

    override fun getPossibleValues(): Collection<Int> = possibleValues

    override fun equals(other: Any?): Boolean {
        if (other == null) {
            return false
        }

        if (this === other) {
            return true
        }

        if (other is FilteredIntegerProperty) {
            if (super.equals(other)) {
                return possibleValues == other.possibleValues
            }
        }

        return false
    }

    override fun getValue(value: String): Optional<Int> {
        try {
            val integer = value.toInt()
            return if (possibleValues.contains(integer)) Optional.of(integer) else Optional.empty()
        } catch (exc: NumberFormatException) {
            return Optional.empty()
        }
    }

    override fun getName(value: Int): String = value.toString()

    override fun generateHashCode(): Int {
        var result = super.generateHashCode()
        result = 31 * result + possibleValues.hashCode()
        return result
    }

    companion object {
        fun create(name: String, possibleValues: Collection<Int>): FilteredIntegerProperty =
            FilteredIntegerProperty(name, possibleValues)

        fun create(name: String, vararg possibleValues: Int): FilteredIntegerProperty =
            FilteredIntegerProperty(name, *possibleValues)
    }
}